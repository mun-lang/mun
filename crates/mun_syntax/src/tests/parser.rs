use crate::SourceFile;

fn ok_snapshot_test(text: &str) {
    let text = text.trim().replace("\n    ", "\n");
    let file = SourceFile::parse(&text);
    let errors = file.errors();
    assert_eq!(
        &*errors,
        &[] as &[crate::SyntaxError],
        "There should be no errors\nAST:\n{}",
        file.debug_dump()
    );
    insta::assert_snapshot!(insta::_macro_support::AutoName, file.debug_dump(), &text);
}

#[test]
fn empty() {
    ok_snapshot_test(r#""#);
}

#[test]
fn function() {
    ok_snapshot_test(
        r#"
    // Source file comment

    // Comment that belongs to the function
    fn a() {}
    fn b(value:number) {}
    export fn c():never {}
    fn b(value:number):number {}"#,
    );
}

#[test]
fn block() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a;
        let b:i32;
        let c:string;
    }"#,
    );
}

#[test]
fn literals() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = true;
        let b = false;
        let c = 1;
        let d = 1.12;
        let e = "Hello, world!"
    }
    "#,
    );
}

#[test]
fn unary_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = --3;
        let b = !!true;
    }
    "#,
    )
}

#[test]
fn binary_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = 3+4*5
        let b = 3*4+10/2
    }
    "#,
    )
}

#[test]
fn expression_statement() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = "hello"
        let b = "world"
        let c
        b = "Hello, world!"
        !-5+2*(a+b);
        -3
    }
    "#,
    )
}

#[test]
fn function_calls() {
    ok_snapshot_test(
        r#"
    fn bar(i:number) { }
    fn foo(i:number) {
      bar(i+1)
    }
    "#,
    )
}

#[test]
fn patterns() {
    ok_snapshot_test(
        r#"
    fn main(_:number) {
       let a = 0;
       let _ = a;
    }
    "#,
    )
}

#[test]
fn compare_operands() {
    ok_snapshot_test(
        r#"
    fn main() {
        let _ = a==b;
        let _ = a == b;
        let _ = a != b;
        let _ = a < b;
        let _ = a > b;
        let _ = a <= b;
        let _ = a >= b;
    }
    "#,
    )
}

#[test]
fn if_expr() {
    ok_snapshot_test(
        r#"
    fn bar() {
        if true {};
        if true {} else {};
        if true {} else if false {} else {};
        if {true} {} else {}
    }
    "#,
    );
}

#[test]
fn block_expr() {
    ok_snapshot_test(
        r#"
    fn bar() {
        {3}
    }
    "#,
    );
}

#[test]
fn return_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        return;
        return 50;
    }
    "#,
    )
}

#[test]
fn loop_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        loop {}
    }"#,
    )
}
