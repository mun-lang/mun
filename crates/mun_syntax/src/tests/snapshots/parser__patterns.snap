---
source: crates/mun_syntax/src/tests/parser.rs
expression: "fn main(_:number) {\n   let a = 0;\n   let _ = a;\n}"
---
SOURCE_FILE@[0; 49)
  FUNCTION_DEF@[0; 49)
    FN_KW@[0; 2) "fn"
    WHITESPACE@[2; 3) " "
    NAME@[3; 7)
      IDENT@[3; 7) "main"
    PARAM_LIST@[7; 17)
      L_PAREN@[7; 8) "("
      PARAM@[8; 16)
        PLACEHOLDER_PAT@[8; 9)
          UNDERSCORE@[8; 9) "_"
        COLON@[9; 10) ":"
        PATH_TYPE@[10; 16)
          PATH@[10; 16)
            PATH_SEGMENT@[10; 16)
              NAME_REF@[10; 16)
                IDENT@[10; 16) "number"
      R_PAREN@[16; 17) ")"
    WHITESPACE@[17; 18) " "
    BLOCK_EXPR@[18; 49)
      L_CURLY@[18; 19) "{"
      WHITESPACE@[19; 23) "\n   "
      LET_STMT@[23; 33)
        LET_KW@[23; 26) "let"
        WHITESPACE@[26; 27) " "
        BIND_PAT@[27; 28)
          NAME@[27; 28)
            IDENT@[27; 28) "a"
        WHITESPACE@[28; 29) " "
        EQ@[29; 30) "="
        WHITESPACE@[30; 31) " "
        LITERAL@[31; 32)
          INT_NUMBER@[31; 32) "0"
        SEMI@[32; 33) ";"
      WHITESPACE@[33; 37) "\n   "
      LET_STMT@[37; 47)
        LET_KW@[37; 40) "let"
        WHITESPACE@[40; 41) " "
        PLACEHOLDER_PAT@[41; 42)
          UNDERSCORE@[41; 42) "_"
        WHITESPACE@[42; 43) " "
        EQ@[43; 44) "="
        WHITESPACE@[44; 45) " "
        PATH_EXPR@[45; 46)
          PATH@[45; 46)
            PATH_SEGMENT@[45; 46)
              NAME_REF@[45; 46)
                IDENT@[45; 46) "a"
        SEMI@[46; 47) ";"
      WHITESPACE@[47; 48) "\n"
      R_CURLY@[48; 49) "}"

